evalml/pipelines/components/transformers/encoders/onehot_encoder.py

```diff
@@ -112,7 +121,7 @@ def fit(self, X, y=None):
         self._encoder = SKOneHotEncoder(categories=categories,
                                         drop=self.parameters['drop'],
                                         handle_unknown=self.parameters['handle_unknown'])
-        self._encoder.fit(X_t[self._cols_to_encode])
+        self._encoder.fit(X_t[self.features_to_encode])
```

Comment:
Right now the encoder is being fit on different data than what it would transform. In transform, we filter to the columns that are categorical. This creates a bug:

```python
from evalml.pipelines import OneHotEncoder
import pandas as pd
df = pd.DataFrame({"a": [1, 2, 3], "b": ["c", "d", "c"]})
ohe = OneHotEncoder(features_to_encode=["a", "b"])
ohe.fit_transform(df)
```

```python
ValueError: The number of features in X is different to the number of features of the fitted data. The fitted data had 2 features and the X has 1 features.
```

I think there are two solutions:
1. Only fit on columns that are in `self.features_to_encode` and categorical
2. Fit/transform on whatever the user passes in as `features_to_encode` if not None. 

I prefer 2 (see my comment in one of your tests as to why)

Reply:
Ah, good catch! Let me update this to encode `features_to_encode` even if they're numeric. I think that makes sense too given that if a user is explicitly passing us the column name then they want it to be one-hot encoded even if its numeric. (arguably, maybe that user should be setting the dtype to categorical / object? :P)

Reply:
Yea I agree that the best solution is to let the user tell us by typing their columns appropriately but that's hairy with pandas. Hopefully #1229 can help with this!